### PR TITLE
feat(WebhookClient): allow creation of clients via URLs

### LINKED ISSUE #6164: WebhookClient creation with url
**Is your feature request related to a problem? Please describe.**
Having to provide the webhook id and token instead of just an url can result in more and less clean code which can be quite annoying.

**Describe the ideal solution**
A method overload that allows users to pass the whole webhook url:
```js
  const webhookClient = new Discord.WebhookClient("https://discord.com/api/webhooks/12334567890/asddgsjhjhshfsjhefhshjfhsjfjehsjhf_ashdjhadjhwhja");
```

**Describe alternatives you've considered**
Using regex to filter out the id and token

**Additional context**
grafik

### PR DESCRIPTION
**Pull request by J-Human**
**Please describe the changes this PR makes and why it should be merged:**

This allows `WebhookClient`'s to be made with URLs, without the need of removing it or doing any type of parsing. This is useful for pass-through services like Webhook senders, etc., allowing a more convenient time for developers who don't want to use IDs and tokens. Example:
```ts
import { WebhookClient } from 'discord.js';

const webhook = new WebhookClient({ url: 'https://discord.com/api/webhooks/1234567890/abc' });

webhook.send({ content: 'Hi.' });
```

This PR also updates:
* Typings.
* Errors (URLs are passed to a RegEx and validated, although 100% perfection is not possible).
* Tests.

It now mandates initializing Webhook clients should have *one* object parameter, that can contain either ID + token or just a URL (this isn't validated so all 3 are optional).

If I do anything wrong, please forgive me, or if this isn't needed, feel free to give the right feedback/review, I appreciate it.

Fixes #6164

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: src/client/WebhookClient.js [hunk 1]
```diff
@@ -10,19 +11,40 @@ const Webhook = require('../structures/Webhook');
...
     super(options);
     Object.defineProperty(this, 'client', { value: this });
-    this.id = id;
-    Object.defineProperty(this, 'token', { value: token, writable: true, configurable: true });
+    let id, token;
+
+    if ('url' in data) {
+      const url = data.url.match(
+        // eslint-disable-next-line no-useless-escape
+        /^https?:\/\/(?:discord|discordapp)\.com\/api\/webhooks(?:\/v[0-9]\d*)?\/([^\/]+)\/([^\/]+)/i,
```
Comment (vladfrangu) [inline src/client/WebhookClient.js:37]: This regexp doesn't account for subdomains. You could also split by slashes

File: src/client/WebhookClient.js [hunk 2]
```diff
@@ -10,19 +11,40 @@ const Webhook = require('../structures/Webhook');
...
+      );
+
+      if (!url || url.length <= 1) throw new Error('WEBHOOK_URL_INVALID');
+
+      id = url[1];
+      token = url[2];
+    }
+
+    this.id = id ?? data.id;
+    Object.defineProperty(this, 'token', { value: token ?? data.id, writable: true, configurable: true });
```
Comment (vladfrangu) [inline src/client/WebhookClient.js:47]: Did you mean token?
```suggestion
    Object.defineProperty(this, 'token', { value: token ?? data.token, writable: true, configurable: true });
```

File: src/errors/Messages.js [hunk 3]
```diff
@@ -101,6 +101,7 @@ const Messages = {
 
   WEBHOOK_MESSAGE: 'The message was not sent by a webhook.',
   WEBHOOK_TOKEN_UNAVAILABLE: 'This action requires a webhook token, but none is available.',
+  WEBHOOK_URL_INVALID: 'The provided webhook URL is not well formed.',
```
Comment (vladfrangu) [inline src/errors/Messages.js:104]: Up to debate if this is better

```suggestion
  WEBHOOK_URL_INVALID: 'The provided webhook URL is not valid.',
```

File: typings/index.d.ts [hunk 4]
```diff
@@ -4340,6 +4340,12 @@ export type VerificationLevel = keyof typeof VerificationLevels;
 
 export type VoiceBasedChannelTypes = 'GUILD_VOICE' | 'GUILD_STAGE_VOICE';
 
+export interface WebhookClientData {
```
Comment (vladfrangu) [inline typings/index.d.ts:4343]: This should be a union of two interfaces that define what props they should have (WebhookClientData = WebhookClientDataURL | WebhookClientDataIdWithToken)

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Comment by J-Human** [comment]
@vladfrangu fixed! Thanks for reminding. 😅

**Review by vladfrangu — APPROVED** [review]

File: src/client/WebhookClient.js [hunk 5]
```diff
@@ -10,19 +11,51 @@ const Webhook = require('../structures/Webhook');
...
     super(options);
     Object.defineProperty(this, 'client', { value: this });
-    this.id = id;
-    Object.defineProperty(this, 'token', { value: token, writable: true, configurable: true });
+    let id, token;
+
+    if ('url' in data) {
+      const url = data.url.match(
+        // eslint-disable-next-line no-useless-escape
+        /^https?:\/\/(?:canary|ptb)?\.?(?:discord|discordapp)\.com\/api\/webhooks(?:\/v[0-9]\d*)?\/([^\/]+)\/([^\/]+)/i,
```
Comment (Jiralite) [inline src/client/WebhookClient.js:48]: I'm not sure if we should be allowing domains with discordapp.com still? Webhooks links generated now across the release channels won't be using "discordapp", and I feel if people are still using that, they should update their webhook url instead?

Comment (iCrawl) [reply src/client/WebhookClient.js:48]: Yeah we can probably remove the "app" part.

**Review by SpaceEEC — APPROVED** [review]

File: src/client/WebhookClient.js [hunk 6]
```diff
@@ -10,19 +11,51 @@ const Webhook = require('../structures/Webhook');
...
+      );
+
+      if (!url || url.length <= 1) throw new Error('WEBHOOK_URL_INVALID');
+
+      id = url[1];
```
Comment (NotSugden) [inline src/client/WebhookClient.js:53]: i think theres a thing for this
`[, id, token] = url;`? something like that

File: src/client/WebhookClient.js [hunk 7]
```diff
@@ -10,19 +11,51 @@ const Webhook = require('../structures/Webhook');
...
+   * const hook = new Discord.WebhookClient({ id: '1234', token: 'abcdef' });
    * hook.send('This will send a message').catch(console.error);
    */
-  constructor(id, token, options) {
+  constructor(data, options) {
     super(options);
     Object.defineProperty(this, 'client', { value: this });
-    this.id = id;
-    Object.defineProperty(this, 'token', { value: token, writable: true, configurable: true });
+    let id, token;
```
Comment (NotSugden) [inline src/client/WebhookClient.js:43]:
```suggestion
    let { id, token } = data;
```
and remove the `id ?? data.id` below (same for token)

**Review by NotSugden — CHANGES_REQUESTED** [review]
the constructor typings need updating still

**Comment by J-Human** [comment]
@NotSugden @Jiralite done! Your suggestions have been committed. Thanks a lot.
* According to Jiralite's comment (and Crawl's decision) - the `discordapp` domain has vanished.
* As per Sugden's comment here, I updated the constructor typings for `WebhookClient` (yes, I forgot to do so), and retrieving the data from either the URL or ID, token has been refactored.

File: src/client/WebhookClient.js [hunk 8]
```diff
@@ -10,17 +11,48 @@ const Webhook = require('../structures/Webhook');
  */
 class WebhookClient extends BaseClient {
   /**
-   * @param {Snowflake} id The webhook's id
-   * @param {string} token Token of the webhook
+   * The data for the webhook client containing either an id and token or just a URL
+   * @typedef {Object} WebhookClientData
+   * @property {(WebhookClientDataIdWithToken|WebhookClientDataURL)} data The data for the webhook client
+   */
```
Comment (NotSugden) [inline src/client/WebhookClient.js:17]: this typedef is
```ts
{
  data: {
    id: string;
    token: string;
  } | {
    url: string;
  }
}
```
convert the typedefs below into a single typedef, and use that instead
(the jsdoc typedefs aren't usually handled this way, the typings are left for the more specific unions)

File: src/client/WebhookClient.js [hunk 9]
```diff
@@ -10,17 +11,48 @@ const Webhook = require('../structures/Webhook');
...
     Object.defineProperty(this, 'client', { value: this });
+    let { id, token } = data;
+
+    if ('url' in data) {
+      const url = data.url.match(
+        // eslint-disable-next-line no-useless-escape
+        /^https?:\/\/(?:canary|ptb)?\.?discord\.com\/api\/webhooks(?:\/v[0-9]\d*)?\/([^\/]+)\/([^\/]+)/i,
+      );
+
+      if (!url || url.length <= 1) throw new Error('WEBHOOK_URL_INVALID');
```
Comment (NotSugden) [inline src/client/WebhookClient.js:51]:
```suggestion
      if (url?.length !== 2) throw new Error('WEBHOOK_URL_INVALID');
```
```suggestion
      if (url?.length < 2) throw new Error('WEBHOOK_URL_INVALID');
```
either or
can also stay the same prob doesn't matter much

**Comment by J-Human** [comment]
Done.

**Review by iCrawl — APPROVED** [review]
Love it, I am actually in need of this too :D

**Review by SpaceEEC — APPROVED** [review]

**Comment by AlexanderGipp** [comment]
Thanks for the addition @J-Human

**Pull request merged**